File: Math and Geometry/Python/Integer to English Words.py
# Problem link: https://leetcode.com/problems/integer-to-english-words/description
# Time: O(n), where n is the number of digits
# Space: O(1)

class Solution:
    def numberToWords(self, num: int) -> str:
        if not num:
            return "Zero"

        between = ["", "Thousand", "Million", "Billion"]
        result, index = [], 0
        store = {1 : "One", 2 : "Two", 3 : "Three", 4 : "Four", 5 : "Five",
                6 : "Six", 7 : "Seven", 8 : "Eight", 9 : "Nine", 10 : "Ten",
                11 : "Eleven", 12 : "Twelve", 13 : "Thirteen", 14 : "Fourteen",
                15 : "Fifteen", 16 : "Sixteen", 17 : "Seventeen", 18 : "Eighteen",
                19 : "Nineteen", 20 : "Twenty", 30 : "Thirty", 40 : "Forty", 50: "Fifty",
                60 : "Sixty", 70 : "Seventy", 80 : "Eighty", 90 : "Ninety"}
        
        def word(num):
            if num in store:
                return store[num]
            
            sen, place = [], 1

            while num > 0:

                if num % 10 != 0:
                    if place == 100:
                        sen.append("Hundred")
                        sen.append(store[num % 10])

                    elif num // 100 > 0 and num % 100 in store:
                        sen.append(store[num % 100])
                        num, place = num // 10, place * 10

                    else:
                        sen.append(store[(num % 10) * place])

                num, place = num // 10, place * 10
            
            return " ".join(sen[::-1])
        

        while num > 0:
            sen = word(num % 1000)

            if sen:
                result.append(between[index])
                result.append(sen)
            
            index += 1
            num //= 1000
        
        return " ".join(result[::-1]).strip()

File: Math and Geometry/Java/Integer to English Words.java
// Problem link: https://leetcode.com/problems/integer-to-english-words/description
// Time: O(n), where n is the number of digits in the given input
// Space: O(1)

class Solution {
    public String numberToWords(int num) {
        if(num == 0) {
            return "Zero";
        }
        
        Map<Integer, String> mapping = new HashMap<>();
        StringBuilder result = new StringBuilder();
        int number = 1, index = 0;


        String[] store = {"", "Thousand", "Million", "Billion", "One", "Two", 
                        "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
                        "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", 
                        "Seventeen", "Eighteen", "Nineteen", "Twenty", "Thirty", "Forty", 
                        "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"};


        for(int idx = 4; idx < store.length; idx++) { 
            mapping.put(number, store[idx]);
            number += (idx >= 23 ? 10 : 1);
        }


        while(num > 0) {
            String sen = this.generate(num % 1000, mapping);

            if(sen.length() > 0)
                result.insert(0, sen + " " + store[index] + " ");

            num /= 1000;
            index++;
        }

        return result.toString().trim();
    }


    private String generate(int num, Map<Integer, String> mapping) {
        if(mapping.containsKey(num)) {
            return mapping.get(num);
        }

        int place = 1;
        StringBuilder sen = new StringBuilder();

        while(num > 0) {
            if(num % 10 != 0) {
                if(place == 100) {
                    sen.insert(0, mapping.get(num % 10) + " Hundred");
                }
                else if(num / 100 > 0 && mapping.containsKey(num % 100)) {
                    sen.insert(0, " " + mapping.get(num % 100));
                    place *= 10;
                    num /= 10;
                }
                else {
                    sen.insert(0, " " + mapping.get(num % 10 * place));
                }
            }

            place *= 10;
            num /= 10;
        }

        return sen.toString().trim();
    }
}
